### ISSUE TITLE
ntpd 'adjtime failed: Invalid argument'

### ISSUE THREAD

**Issue by teto** [NixOS/nixpkgs#31885]
## Issue description
I have openntpd configured as
```
services.openntpd = {
    enable = true;
}
```
my clock is off by too much for ntpd to correct
```
nov. 21 19:30:55 jedha systemd[1]: Starting OpenNTP Server...
nov. 21 19:30:55 jedha systemd[1]: Started OpenNTP Server.
nov. 21 19:32:26 jedha ntpd[770]: adjusting local clock by -32394.968842s
nov. 21 19:32:26 jedha ntpd[770]: adjtime failed: Invalid argument
```
so I tried to run `ntpd -s -d` which seems like the standard way to force a clock update                                                                                                                                                 
`/etc/ntpd.conf: No such file or directory`

Looking at the derivation, there seems to be 2 conflicting sysconfdir @wkennington ?
```
  configureFlags = [
    "--with-privsep-path=${privsepPath}"
    "--with-privsep-user=${privsepUser}"
    "--sysconfdir=/etc"
    "--localstatedir=/var"
    "--with-cacert=/etc/ssl/certs/ca-certificates.crt"
  ];
...
  installFlags = [
    "sysconfdir=\${out}/etc"
    "localstatedir=\${TMPDIR}"
  ];
```

## Technical details

 - system: `"x86_64-linux"`
 - host os: `Linux 4.14.0, NixOS, 18.03.git.b8f7027 (Impala)`
 - multi-user?: `yes`
 - sandbox: `no`
 - version: `nix-env (Nix) 1.11.15`
 - channels(teto): `"nixos-unstable-18.03pre120398.aeff3080d0, nixos-17.09.2075.ac35504065"`
 - channels(root): `"nixos-17.09.2075.ac35504065"`
 - nixpkgs: `/home/teto/nixpkgs`

**Comment by teto** [comment]
```
diff --git a/pkgs/tools/networking/openntpd/default.nix b/pkgs/tools/networking/openntpd/default.nix
index 7e1c257ec4b..a1630db3976 100644
--- a/pkgs/tools/networking/openntpd/default.nix
+++ b/pkgs/tools/networking/openntpd/default.nix
@@ -15,7 +15,7 @@ stdenv.mkDerivation rec {
   configureFlags = [
     "--with-privsep-path=${privsepPath}"
     "--with-privsep-user=${privsepUser}"
-    "--sysconfdir=/etc"
+    "--sysconfdir=\${out}/etc"
     "--localstatedir=/var"
     "--with-cacert=/etc/ssl/certs/ca-certificates.crt"
   ];
@@ -23,7 +23,6 @@ stdenv.mkDerivation rec {
   buildInputs = [ libressl ];
 
   installFlags = [
-    "sysconfdir=\${out}/etc"
     "localstatedir=\${TMPDIR}"
   ]
```
seems to fix it

**Comment by teto** [comment]
@wkennington should I open a PR ? What should I do about the duplicated "localstatedir=\${TMPDIR}" ?

**Comment by teto** [comment]
I looked it up again and it seems that the fault lies in the module:
https://github.com/NixOS/nixpkgs/blob/8152db99d75e690b96e091b1b1a50f7325a30981/nixos/modules/services/networking/openntpd.nix#L74
`ExecStart = "${package}/sbin/ntpd -f ${cfgFile} -p ${pidFile} ${cfg.extraOptions}";`
It starts ntpd with a specific config file, but then once you tried to run `ntpd -s` as a user, it tries to read from `/etc/ntpd.conf: No such file or directory` and fails miserably

**Comment by coretemp** [comment]
I had the same problem and switched to timesyncd just to get `date` to return something sane again. I don't know whether this will resolve things long term. 

On the Arch wiki `openntpd adjtime failed: Invalid argument` is also documented as common issue. It is my understanding that with dozens of NTP servers available globally showing the wrong time shouldn't be a major technical challenge.

openntpd doesn't cope with changes larger than 1000 seconds, IIRC, and as such other machinery might have to check via some quorom method in such cases by default. E.g. when asking Google what the current time is would be pretty reliable, given that they use atomic clocks. 

In short, openntpd as it is currently (on 17.09 release channel) being distributed and configured within NixOS is not a reliable method to make `date` return an answer which is reasonable (my clock was about a day in the future). 

If there is a method to keep time which already has the reliability I am looking for, I am all ears. I know Google also runs a free time service which smooths time. I don't know the real world reliability of that over time, but it would be of interest to see data on that too and perhaps switch to it as default. Then, if other providers implement a similar method, there would also be redundancy.

**Comment by teto** [comment]
PR merged => closing. I think ntpd might have the same problem (and same solution)
